src/main/kotlin/me/serce/solidity/ide/formatting/SolidityExternalFormatter.kt

```diff
@@ -0,0 +1,122 @@
+
+  override fun createFormattingTask(request: AsyncFormattingRequest): FormattingTask? {
+    val formattingContext: FormattingContext = request.context
+    val project = formattingContext.project
+    val settings = SoliditySettings.getInstance(project)
+    val userHome = System.getProperty("user.home")
+
+    if (settings.formatterType == FormatterType.INTELLIJ_SOLIDITY) {
+      return null
+    }
+
+    val psiFile: PsiFile = request.context.containingFile
+    val foundryExePath = settings.executablePath.ifBlank { "$userHome/.foundry/bin/forge" }
+
+    return try {
+      val cmd = GeneralCommandLine()
+        .withParentEnvironmentType(GeneralCommandLine.ParentEnvironmentType.CONSOLE)
+        .withWorkDirectory(project.basePath)
+        .withExePath(foundryExePath)
+        .withParameters(buildList {
+          add("fmt")
+          add("-")
+          add("--raw")
+
+          if (settings.configPath.isNotBlank()) {
+            add("--root")
+            add(settings.configPath)
+          }
+        })
+        .withCharset(StandardCharsets.UTF_8)
+
+      val handler = OSProcessHandler(cmd)
+      handler.processInput.use { outputStream ->
+        outputStream.write(psiFile.text.toByteArray())
+        outputStream.flush()
+      }
+      object : FormattingTask {
+        override fun run() {
+          val adapter = object : CapturingProcessAdapter() {
+            override fun processTerminated(event: ProcessEvent) {
+              val exitCode = event.exitCode
+              if (exitCode == 0) {
+                request.onTextReady(output.stdout)
+              } else {
+                val onlyErrors = removeWarning(output.stderr).ifBlank { "Formatting failed with exit code $exitCode" }
+                request.onError("Solidity", onlyErrors)
+              }
+            }
+          }
+          handler.addProcessListener(adapter)
+          handler.startNotify()
+        }
+
+        override fun cancel(): Boolean {
+          handler.destroyProcess()
+          return true
+        }
+
+        override fun isRunUnderProgress(): Boolean = true
+      }
+    } catch (e: ExecutionException) {
+      request.onError("Solidity", e.message ?: "Unknown error")
+      null
+    }
+  }
+
+  private fun isWarning(line: String): Boolean {
+    val l = line.trim().lowercase()
+    return l.startsWith("warning:")
+  }
+
+  private fun removeWarning(stderr: String): String {
+    val lines = stderr.lineSequence().toList()
+    val errors = lines.filter { !isWarning(it) }
+    return when {
+      errors.isNotEmpty() -> errors.joinToString("\n")
+      else -> stderr
+    }
+  }
+
+
+  override fun getNotificationGroupId(): String {
+    return "Solidity"
+  }
+
+  override fun getName(): String {
+    return "Solidity"
+  }
+
+  override fun getFeatures(): Set<Feature?> {
+    return EnumSet.noneOf(Feature::class.java)
+  }
+
+  override fun canFormat(p0: PsiFile): Boolean {
```

Comment:
```suggestion
  override fun canFormat(file: PsiFile): Boolean {
```

I'm guessing the initial method override was done when the source code wan't downloaded, so it couldn't see the actual name.

Reply:
I naively thought that Intellij would handle that by itself. Thank you for the patch!